returnn/tf/layers/basic.py

```diff
@@ -4769,39 +4769,55 @@ def __init__(self, eval, **kwargs):
 
 class CompareLayer(LayerBase):
   """
-  This layer performs an elementwise comparison of all provided sources, and returnns a boolean vector
-  of the same shape.
-  If the `value` parameter is provided, all elements in the provided sources are also compared to this value.
+  Compares element-wise the tokens of all input sequences among themselves and/or with a specified given value.
+  The comparisons are performed in a chain according to the order in which they are listed. 
+
+  Example::
+
+    {"class": "compare", "from": ["i1", "i2"], "value": val, "kind": "less"}
+  computes i1 < i2 < val and it is true only if the whole chain of operations is true.
+  The final result is the logical "and" of all comparisons. Note that `value` is the last element to be compared to.
+  
+  A common example usage is the `end` layer in a rec subnetwork to specify the stopping criterion,
+  e.g. the last generated token is equal to the end-of-sentence token::
+
+    "output": {"class": "rec", "from": [], "unit": {
+        .
+        .
+        .
+        "end": {"class": "compare", "from": "output", "value": end_of_sentence_id}
+    }, "target": "classes0"}
 
-  If more than two sources are provided, each additional source will be compared against the first source,
-  and the results will be combined by a logical conjunction.
   """
   layer_class = "compare"
 
   def __init__(self, kind="equal", value=None, **kwargs):
     """
-    :param str kind: Which comparison operation to use, e.g. "equal" or other supported tf comparison ops
+    :param str kind: which comparison operation to use, e.g. "equal", "greater", "less"
+      or other supported TF comparison ops
     :param float|int|None value: if specified, will also compare to this
     """
     super(CompareLayer, self).__init__(**kwargs)
     assert len(self.sources) >= 1
     if value is None:
-      assert len(self.sources) >= 2
+      assert len(self.sources) >= 2, "{} requires at least two elements to compare".format(self)
     op = getattr(tf, kind)  # e.g. tf.equal
     from returnn.tf.util.basic import swapaxes
-    x = self.sources[0].output.placeholder
+    common_data = Data.get_common_data([s.output for s in self.sources])
+    x = self.sources[0].output.copy_compatible_to(common_data).placeholder
     batch_axis = self.sources[0].output.batch_dim_axis
     r_last = None
-    if value is not None:
-      r_last = op(x, value)
     for source in self.sources[1:]:
-      x2 = source.output.placeholder
+      x2 = source.output.copy_compatible_to(common_data).placeholder
       if source.output.batch_dim_axis != batch_axis:
         x2 = swapaxes(x2, batch_axis, source.output.batch_dim_axis)
       r = op(x, x2)
       if r_last is not None:
         r = tf.logical_and(r_last, r)
       r_last = r
+      x = x2
+    if value is not None:
+      r_last = tf.logical_and(r_last, op(x, value))
```

Comment:
```suggestion
      r_last = opt_logical_and(r_last, op(x, value))
```

Reply:
There was still the problem that `r_last` could have been `None` here.
But I added a few suggestions to use `opt_logical_and`, which simplifies the code anyway.